addons/account/models/account_move.py

```diff
@@ -1731,10 +1731,10 @@ def create_tax_cash_basis_entry(self, percentage_before_rec):
                 for line in move.line_ids:
                     if not line.tax_exigible:
                         #amount is the current cash_basis amount minus the one before the reconciliation
-                        if percentage_after == 1.0 and line.amount_residual:
+                        amount = line.balance * percentage_after - line.balance * percentage_before
+                        if percentage_after == 1.0 and percentage_before > 0.0 and line.amount_residual < 0 and amount < 0 and float_compare(line.amount_residual, amount, precision_rounding=line.company_id.currency_id.rounding) < 0:
```

Comment:
I don't understand the condition. 
why is it not 
```python
if percentage_after:
    # when registering the final payment we use directly the amount_residual to correct rounding issues
    amount = line.amount_residual
else:
    amount = line.balance * percentage_after - line.balance * percentage_before
    amount = line.company_id.currency_id.round(amount)

...
```
?

Reply:
*tl;dr* we pass though this point when sending/receiving payments, so percentage_after is > 0 but amount_residual have different meaning depending on the exact line/direction of the payment

long version:
When registering a payment, the flow go through
`_create_payment_entry -> register_payment -> reconcile -> auto_reconcile_lines -> _reconcile_lines -> create_tax_cash_basis_entry`
here we process usually 3 lines, 2 debit lines (product and tax) and 1 credit line, the payment

I'll report just the relevant data of the lines before https://github.com/odoo/odoo/pull/63738/files#diff-1e3bd6be3bfb83a37ec9fb800ce8b1c95afe0be90ff792874ae7299c320a2f6eL1734
**for this use case** (registering a full payment at once)
```
tax line
percentage_after 1.0
percentage_before 0.0
amount_residual 0.0
balance 2095.3

product line
percentage_after 1.0
percentage_before 0.0 
amount_residual 0.05
balance 13095.65
```

The correct amount for both lines is `line.balance * percentage_after - line.balance * percentage_before` as the `amount_residual` take into consideration the unpaid amount

**Consider now a different scenario** from e25f857e672a645bfc46a9bac9b871c107b0a8db (same tax settings).
We receive several payments (so incoming now) for an invoice, in the last one we have

```
tax line
percentage after 1.0
percentage before 0.73333333
line.amount_residual -13.34
line.balance -50.0

product line
percentage after 1.0
percentage before 0.73333333
line.amount_residual 0.0
line.balance -250.0
```

The amount would be
`line.balance * percentage_after - line.balance * percentage_before` -> `-13.333333333333336` -> `-13.33` after rounding. 
A 0.01 will be missing, lost in the computation

Reply:
I read you but I don't see how it answers my question: why is it not 
``` python
if percentage_after:
    # when registering the final payment we use directly the amount_residual to correct rounding issues
    amount = line.amount_residual
else:
    amount = line.balance * percentage_after - line.balance * percentage_before
    amount = line.company_id.currency_id.round(amount)

```
?

In your second example, you'd be using the amount_residual of the tax line (13.34) instead of recomputing it using `line.balance * percentage_after - line.balance * percentage_before`

Reply:
Because we would always take as amount `line.amount_residual` (`percentage_after` is always `> 0`) and never recompute using the formula

Analyzing data for the caba entry of this PR use case:
```
#  only the product line to simplify
percentage_after 1.0
percentage_before 0.0 
amount_residual 0.05
balance 13095.65
```
With your suggestion `amount = line.amount_residual = 0.05` and the caba entry would look like
![caba-before](https://user-images.githubusercontent.com/53174659/106126516-bdda8680-615d-11eb-93e0-2d336a5040b8.png)

As you can see it's wrong, it should be `amount = line.balance * percentage_after - line.balance * percentage_before =  13095.65`

![caba-2](https://user-images.githubusercontent.com/53174659/106126599-d77bce00-615d-11eb-9de3-f48e2c022526.png)

Reply:
mmh ok, thanks for hte explanation :+1: